ext/libxml/libxml.c

```diff
@@ -1071,29 +1063,44 @@ PHP_FUNCTION(libxml_disable_entity_loader)
 /* {{{ Changes the default external entity loader */
 PHP_FUNCTION(libxml_set_external_entity_loader)
 {
+	zval					*callback;
 	zend_fcall_info			fci;
 	zend_fcall_info_cache	fcc;
+	char					*error = NULL;
 
 	ZEND_PARSE_PARAMETERS_START(1, 1)
-		Z_PARAM_FUNC_OR_NULL(fci, fcc)
+		Z_PARAM_ZVAL(callback)
```

Comment:
what was the problem with the original ZPP? We put a lot of effort into using ZPP anywhere where it's possible (in order to make param parsing surely consistent), so I'd be happy if we could keep this one.

Reply:
We need the zval so that we can save it and later return it from a call to libxml_get_external_entity_loader().

Reply:
I see, but the original implementation created and then also stored the zval as the `object` member, didn't it?

Reply:
It saved fcc->object, but that is not necessarily the same thing as the input parameter, see zend_is_callable_at_frame(). If the input was a string, or an array with two strings (a static call), fcc->object will be NULL. It would be awkward to try to put the input zval back together after zend_is_callable_at_frame() has pulled it apart.

Reply:
I think the best alternative is to introduce a facility for this use case into zend_API.h. Two options occur to me. An extremely specific macro:

```
	ZEND_PARSE_PARAMETERS_START(1, 1)
		Z_PARAM_CALLABLE_OR_NULL_WITH_ZVAL(fcc, fci, z_callback)
	ZEND_PARSE_PARAMETERS_END();
```

Or a more generic solution. After the argument is parsed, `_arg` still points to the zval, so it's possible to capture it at that point:

```
	ZEND_PARSE_PARAMETERS_START(1, 1)
		Z_PARAM_CALLABLE_OR_NULL(fcc, fci)
		Z_PARAM_GET_PREV_ZVAL(z_callback)
	ZEND_PARSE_PARAMETERS_END();
```

Reply:
I pushed the second solution for your consideration. It does look quite neat. Note that ugly bits of the LHS like _php_libxml_destroy_fci() have gone away because the zval keeps everything in fci/fcc alive.

Reply:
I think the generic solution is probably better here as it could be used elsewhere and looks quite simple.

Reply:
TBH I like the specific one more because parameter counting stays easy, as we don't have to discount the `Z_PARAM_GET_PREV_ZVAL` lines.